src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs

```diff
@@ -0,0 +1,371 @@
+
+pub enum HookType {
+    PreStart,
+    CreateRuntime,
+    PostStart,
+    PostStop,
+}
+
+// execute hooks in oci::Hooks according to hook types
+pub async fn run_hooks(spec: &oci::Spec, st: &oci::State, hk_type: HookType) -> Result<()> {
+    if let Some(spec_hooks) = &spec.hooks {
+        let hooks = match hk_type {
+            HookType::PreStart => {
+                if !spec_hooks.prestart.is_empty() {
+                    info!(sl!(), "prestart hook");
+                }
+                &spec_hooks.prestart
+            }
+            HookType::CreateRuntime => {
+                if !spec_hooks.create_runtime.is_empty() {
+                    info!(sl!(), "create runtime hook");
+                }
+                &spec_hooks.create_runtime
+            }
+            HookType::PostStart => {
+                if !spec_hooks.poststart.is_empty() {
+                    info!(sl!(), "poststart hook");
+                }
+                &spec_hooks.poststart
+            }
+            HookType::PostStop => {
+                if !spec_hooks.poststop.is_empty() {
+                    info!(sl!(), "poststop hook");
+                }
+                &spec_hooks.poststop
+            }
+        };
+
+        for h in hooks.iter() {
+            run_hook(h, st).await?;
+        }
+    }
+
+    Ok(())
+}
+
+// execute a specific hook, regardless of hook type
+pub async fn run_hook(h: &oci::Hook, st: &oci::State) -> Result<()> {
+    info!(sl!(), "runtime action: execute hook");
+
+    // check whether the binary file called by the hook exists
+    let binary_path = PathBuf::from(h.path.as_str()).canonicalize()?;
+    if !binary_path.exists() {
+        return Err(anyhow!("Hook Binary path {:?} does not exist", binary_path));
+    }
+
+    // hook.args[0] is the binary name, which should not be included in Command.args
+    let mut command_args = h.args.clone();
+    if !command_args.is_empty() {
+        command_args.remove(0);
+    }
+
+    // split envs, and those invalid envs will be omitted
+    let command_envs: HashMap<&str, &str> = h.env.iter().filter_map(|e| valid_env(e)).collect();
+
+    // set execution timeout duration
+    let execution_timeout = match h.timeout {
+        Some(t) => Duration::new(t as u64, 0),
+        None => Duration::MAX,
+    };
+
+    // create tokio child process to execute hook
+    let mut child = tokio::process::Command::new(&binary_path)
+        .args(command_args.iter())
+        .envs(command_envs.iter())
+        .kill_on_drop(true)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let state = serde_json::to_string(st)?;
+    let join_handle = tokio::spawn(async move {
+        // write state to stdin of child process
+        if let Some(mut stdin) = child.stdin.take() {
+            match stdin.write_all(state.as_bytes()).await {
+                Ok(_) => {}
+                Err(e) => {
```

Comment:
Should this return an error?

Reply:
We probably don't need to return error here. I think it is best not to hinder the continued operation of the runtime because of the execution of the hook. But the log type here may indeed need to be changed.